src/nfts/seasons/ERC1155SeasonOne.sol

```diff
@@ -0,0 +1,82 @@
+import {SeasonsBase} from "@protocol/nfts/seasons/SeasonsBase.sol";
+import {CoreRef} from "@protocol/refs/CoreRef.sol";
+import {Roles} from "@protocol/core/Roles.sol";
+import {ERC1155MaxSupplyMintable} from "@protocol/nfts/ERC1155MaxSupplyMintable.sol";
+
+contract ERC1155SeasonOne is SeasonsBase {
+    using SafeERC20 for IERC20;
+
+    mapping(uint256 tokenId => uint256 rewardAmount) public tokenIdRewardAmount;
+    mapping(uint256 tokenId => uint256 usedAmount) public tokenIdUsedAmount;
+
+    constructor(
+        address _core,
+        address _nftContract,
+        address _rewardToken,
+        address _seasonsTokenIdRegistryContract
+    ) SeasonsBase(_core, _nftContract, _rewardToken, _seasonsTokenIdRegistryContract) {}
+
+    /// @notice config the season distribution
+    /// @dev _nftContract much have setSupplyCap() set or this tx will revert.
+    function configSeasonDistribution(
+        TokenIdRewardAmount[] memory tokenIdRewardAmounts
+    ) external override onlyRole(Roles.DEPLOYER) sealAfter returns (uint256) {
+        uint256 _totalRewardTokens = 0;
+        // set tokenIdRewardAmount mapping
+        for (uint256 i = 0; i < tokenIdRewardAmounts.length; i++) {
+            require(tokenIdRewardAmounts[i].rewardAmount > 0, "ERC1155SeasonOne: rewardAmount cannot be 0");
+            tokenIdRewardAmount[tokenIdRewardAmounts[i].tokenId] = tokenIdRewardAmounts[i].rewardAmount;
+
+            uint _maxTokenSupply = ERC1155MaxSupplyMintable(nftContract).maxTokenSupply(
+                tokenIdRewardAmounts[i].tokenId
+            );
+            require(_maxTokenSupply != 0, "ERC1155SeasonOne: maxTokenSupply cannot be 0");
+
+            // Running total of totalPaymentToken amount needed by the contract to be solvent
+            _totalRewardTokens += (tokenIdRewardAmounts[i].rewardAmount * _maxTokenSupply);
+        }
+
+        registerTokenIds(tokenIdRewardAmounts);
+
+        totalRewardTokens = _totalRewardTokens;
+        return totalRewardTokens;
+    }
+
+    function redeem(uint256 tokenId) public override {
+        /// ---- checks ---- ///
+        require(solvent(), "ERC1155SeasonOne: Contract Not solvent");
+        require(tokenIdRewardAmount[tokenId] > 0, "ERC1155SeasonOne: No redeemable tokens for given tokenId");
+        require(
+            ERC1155MaxSupplyMintable(nftContract).balanceOf(msg.sender, tokenId) > 0,
+            "ERC1155SeasonOne: No capsule available in users wallet"
+        );
+
+        /// ---- effects ---- ///
+        uint256 _rewardAmount = tokenIdRewardAmount[tokenId];
+
+        // Increase the counter for the amount of reward tokens used
+        tokenIdUsedAmount[tokenId] += _rewardAmount;
+
+        // Update total reward tokens needed in contract
+        totalRewardTokens -= _rewardAmount;
+
+        /// ---- interaction ---- ///
+        // Transfer
+        nftContract.safeTransferFrom(msg.sender, address(this), tokenId, 1, "");
```

Comment:
Need to discuss this further, as `msg.sender` won't be the recipient in all cases (especially with account abstraction). We might need a hash/sig check similar to what we've done in `GameConsumer.sol` or `ERC1155AutoGraphMinter.sol`.

Reply:
We shouldnt need a hash as the client has to approve the sending in of the NFT from whatever wallet it is getting sent from. So as long as there is a way to approve the sending of the NFT from said wallet address into the contract that is all thats needed. Refer to the test for the approve function details.

Reply:
@utx0 - we'll take this offline as there's more context/nuance to chat about.